src/org/sipdroid/sipua/PublishAgent.java

```
@@ -0,0 +1,215 @@
+import org.zoolu.sip.header.ProxyAuthorizationHeader;
+import org.zoolu.sip.header.StatusLine;
+import org.zoolu.sip.header.ViaHeader;
+import org.zoolu.sip.header.WwwAuthenticateHeader;
+import org.zoolu.sip.message.Message;
+import org.zoolu.sip.message.MessageFactory;
+import org.zoolu.sip.provider.SipProvider;
+import org.zoolu.sip.provider.SipStack;
+import org.zoolu.sip.transaction.TransactionClient;
+import org.zoolu.sip.transaction.TransactionClientListener;
+import org.zoolu.tools.LogLevel;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.logging.Logger;
+
+
+public class PublishAgent implements TransactionClientListener {
+
+	/**
+	 * UserProfile
+	 */
+	protected UserAgentProfile user_profile;
+	/**
+	 * User name.
+	 */
+	String username;
+	/**
+	 * User realm.
+	 */
+	String realm;
+	/**
+	 * Nonce for the next authentication.
+	 */
+	String next_nonce;
+	/**
+	 * User's passwd.
+	 */
+	String passwd;
+	/**
+	 * Expiration time.
+	 */
+	int expire_time;
+
+	private Logger logger = Logger.getLogger(getClass().getCanonicalName());
+	/**
+	 * SipProvider
+	 */
+	protected SipProvider sip_provider;
+	Context context;
+
+	public PublishAgent(SipProvider sip_provider, UserAgentProfile user_profile, String username, String realm, String passwd, Context context) {
+		this.sip_provider = sip_provider;
+		this.user_profile = user_profile;
+		this.username = username;
+		this.realm = realm;
+		this.passwd = passwd;
+		this.next_nonce = null;
+		this.context = context;
+
+	}
+
+	public void publish() {
+		this.expire_time = SipStack.default_expires;
+		this.publish("open", expire_time, "");
+	}
+
+	public void publish(String status, int expireTime, String note) {
+		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+		Boolean publish_enable_status = prefs.getBoolean("publish_enable", true);
+		if (publish_enable_status == true && (status=="open" || status=="close")) {

```

please, throw an exception if `status` is invalid. Silently ignoring invalid values will do no good.